src/windows.rs

```diff
@@ -0,0 +1,343 @@
+/// Dummy value for SIGPROF which is missing in Windows.
+pub const SIGPROF: c_int = DUMMY_OFFSET + 27;
+/// Dummy value for SIGQUIT which is missing in Windows.
+pub const SIGQUIT: c_int = DUMMY_OFFSET + 3;
+/// Dummy value for SIGSTOP which is missing in Windows.
+pub const SIGSTOP: c_int = DUMMY_OFFSET + 19;
+/// Dummy value for SIGSYS which is missing in Windows.
+pub const SIGSYS: c_int = DUMMY_OFFSET + 31;
+/// Dummy value for SIGTRAP which is missing in Windows.
+pub const SIGTRAP: c_int = DUMMY_OFFSET + 5;
+/// Dummy value for SIGUSR1 which is missing in Windows.
+pub const SIGUSR1: c_int = DUMMY_OFFSET + 10;
+/// Dummy value for SIGUSR2 which is missing in Windows.
+pub const SIGUSR2: c_int = DUMMY_OFFSET + 12;
+/// Dummy value for SIGWINCH which is missing in Windows.
+pub const SIGWINCH: c_int = DUMMY_OFFSET + 28;
+
+/// List of forbidden signals.
+///
+/// Some signals are impossible to replace according to POSIX and some are so special that this
+/// library refuses to handle them (eg. SIGSEGV). The routines panic in case registering one of
+/// these signals is attempted.
+///
+/// See [`register`](fn.register.html).
+pub const FORBIDDEN: &[c_int] = &[SIGKILL, SIGSTOP, SIGILL, SIGFPE, SIGSEGV];
+
+#[derive(Copy, Clone, Debug, Eq, PartialEq, Ord, PartialOrd, Hash)]
+struct ActionId(u64);
+
+/// An ID of registered action.
+///
+/// This is returned by all the registration routines and can be used to remove the action later on
+/// with a call to [`unregister`](fn.unregister.html).
+#[derive(Copy, Clone, Debug, Eq, PartialEq, Ord, PartialOrd, Hash)]
+pub struct SigId {
+    signal: c_int,
+    action: ActionId,
+}
+
+type Action = Fn() + Send + Sync;
+
+struct GlobalData {
+    // RCU isn't necessary here. Unlike posix signals, Windows Console handler is called in a separate thread.
+    state: Mutex<GlobalDataState>,
+}
+
+struct GlobalDataState {
+    action_id: u64,
+    all_signals: HashMap<c_int, BTreeMap<ActionId, Box<Action>>>,
+}
+
+impl GlobalData {
+    fn ensure() -> &'static Self {
+        use std::cell::UnsafeCell;
+        use std::sync::Once;
+
+        struct GlobalCell(UnsafeCell<Option<GlobalData>>);
+        unsafe impl Sync for GlobalCell {}
+        static GLOBAL: GlobalCell = GlobalCell(UnsafeCell::new(None));
+        static GLOBAL_ONCE: Once = Once::new();
+
+        GLOBAL_ONCE.call_once(|| unsafe {
+            *GLOBAL.0.get() = Some(GlobalData::new());
+        });
+
+        if let Some(ref data) = *unsafe { &*GLOBAL.0.get() } {
+            data
+        } else {
+            unreachable!();
+        }
+    }
+
+    fn new() -> Self {
+        assert!(unsafe { SetConsoleCtrlHandler(Some(handler), TRUE) != 0 });
+        GlobalData {
+            state: Mutex::new(GlobalDataState::new()),
+        }
+    }
+}
+impl GlobalDataState {
+    fn new() -> Self {
+        Self {
+            action_id: 0,
+            all_signals: HashMap::new(),
+        }
+    }
+}
```

Comment:
This gymnastics is needed on unix systems to make sure only async-signal-safe primitives are used in the signal handlers. I believe it is fine to use something like `std::sync::Once` or `lazy_static` here and get rid of the whole lot of code.

Reply:
Ah, I misunderstood the reason why `lazy_static` isn't used here.